Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm

```diff
@@ -53,17 +57,91 @@
 static NSString * const worldKey = @"world";
 
 static NSString * const  cssKey = @"css";
-// FIXME: <https://webkit.org/b/261765> Consider adding support for cssOrigin.
 
 static NSString * const mainWorld = @"MAIN";
 static NSString * const isolatedWorld = @"ISOLATED";
 
 static int64_t const mainFrameID = 0;
 
+// FIXME: <https://webkit.org/b/261765> Consider adding support for cssOrigin.
 // FIXME: <https://webkit.org/b/261765> Consider adding support for injectImmediately.
 
 namespace WebKit {
 
+static inline NSArray *toAPI(Vector<WebExtensionScriptInjectionResultParameters> results)
+{
+    NSMutableArray *injectionResults = [NSMutableArray arrayWithCapacity:results.size()];
+    for (auto& result : results) {
+        NSDictionary *details = @{
+            @"result": (NSString *)result.result.value(),
+            @"frameId": @(result.frameID.value()),
+        };
+
+        if (result.error)
+            details = mergeDictionariesAndSetValues(details, @{ @"error": result.error.value() });
+
+        [injectionResults addObject:details];
+    }
+
+    return injectionResults;
+}
+
+static inline void parseTargetInjectionOptions(NSDictionary *targetInfo, WebExtensionScriptInjectionParameters& parameters, NSString **outExceptionString)
+{
+    NSNumber *tabId = targetInfo[tabIDKey];
+    std::optional<WebExtensionTabIdentifier> identifier = toWebExtensionTabIdentifier(tabId.doubleValue);
+    if (!identifier) {
+        *outExceptionString = toErrorString(nil, tabIDKey, @"'%@' is not a tab identifier", tabId);
+        return;
+    }
+
+    parameters.identifier = identifier;
+
+    if (NSArray *frameIds = targetInfo[frameIDsKey]) {
+        Vector<size_t> frames;
+        for (NSNumber *frameId in frameIds)
+            frames.append(frameId.unsignedLongLongValue);
```

Comment:
This should really use `toWebExtensionFrameIdentifier()`, and those an exception if any come back as invalid.